### ISSUE TITLE
Lazy tensor loader

### ISSUE THREAD

**Issue by sudy-super** [cg123/mergekit#57]
Adding the "--lazy_unpickle" flag when performing a merge consumes less memory. What is the principle behind this? Also, does it reduce the accuracy?

**Comment by cg123** [comment]
Without `--lazy-unpickle`, mergekit uses `torch.load` to load the shards of the model. This loads the entire shard into memory all at once. When you start involving multiple models this adds up really quickly.

The way `--lazy-unpickle` works is by monkeypatching `torch` and `pickle` so that `torch.load` doesn't actually load any tensors, but instead just returns the locations of the tensors in the file so they can be read as needed. Then during the merge we can load a single tensor at a time.

There should be no decrease in accuracy from this - it should give exactly identical output. There are two reasons I have this gated under a separate flag. Perhaps the biggest one is that it does tinker with the internals with pytorch and I'm inherently leery of making that the standard path. Secondly, there are a lot of weird and different pickled models out there - the lazy unpickler has pretty good coverage but the pickle format is flexible enough that I can't dream of supporting 100% of it.

Hope this helps!

**Comment by sudy-super** [comment]
Thank you for your polite response! Is it possible to apply this loading method to inference?

**Comment by cg123** [comment]
It might be possible to use a method like this for inference as well, but I'm not sure if you would actually see much performance benefit. It works great for mergekit because we really only need to touch each tensor more-or-less once - we read it into memory, perform a merge operation, then almost immediately write it out.

When you're inferencing you need to use each tensor once *per token*. So unless you unload the tensors after use and reload them again for the next token, you wouldn't really get any memory benefit - and if you do unload the tensors like that, it's going to introduce some really heavy IO overhead that'll kill performance in a big way.

So to summarize, yes, but it'd probably hurt more than help. :) Hope this is useful.
